### PR TITLE
[WIP] More flexible dims

### PR DESCRIPTION
**Pull request by JanMarvin**
Still a little hackish solution to the flexibility problem mentioned in #796 , but it shows what is possible. For whatever reason the enforce option does not work with `wb_add_fill()` probably because empty cells are still initialized even though they are not filled.

Also there is still a bug, for now I simply assumed that all data columns are beneath each other to fill their positions. In the example below the column header is right to the data and hence the row has to be filled wider, which is not yet implemented. Hence the first data part is filled incorrectly.

``` r
library(openxlsx2)

# write a workbook with a few colored cells
wb <- wb_workbook()$
  add_worksheet()$
  add_fill(dims = "A1:B2;C3:D4;E1:F2;G5:H6;I2:K2", color = wb_color("yellow"))

# force a dataset into a specific cell dimension
options("openxlsx2.enforce_dims" = TRUE)
wb$add_data(dims = "I2:J2;A1:B2;G5:H6", x = matrix(1:8, 4, 2))
options("openxlsx2.enforce_dims" = FALSE)

if (interactive()) wb$open()
```

Screenshot 2024-04-13 at 13 26 30

### CODE REVIEW

File: tests/testthat/test-wb_functions.R [hunk 1]
```diff
@@ -188,11 +188,12 @@ test_that("dims_to_dataframe", {
...
-  df <- dims_to_dataframe("A1:D5;F1:F6;D8", fill = TRUE)
-  dims <- dataframe_to_dims(df)
-  df2 <- dims_to_dataframe(dims, fill = TRUE)
-  expect_equal(df, df2)
+  ## I should probably fix this failing test
+  # # dims_to_dataframe will always create a square
+  # df <- dims_to_dataframe("A1:D5;F1:F6;D8", fill = TRUE)
+  # dims <- dataframe_to_dims(df)
+  # df2 <- dims_to_dataframe(dims, fill = TRUE)
+  # expect_equal(df, df2)
```
Comment (JanMarvin) [inline tests/testthat/test-wb_functions.R:196]: This test is not so important. It wasn't working entirely correct previously, it simply created two data frames of similar size, but did not care about not selected cells

File: src/helper_functions.cpp [hunk 2]
```diff
@@ -237,6 +237,59 @@ SEXP copy(SEXP x) {
   return Rf_duplicate(x);
 }
 
+// FIXME C++17 is quite a jump for such a tiny function ...
```
Comment (JanMarvin) [inline src/helper_functions.cpp:240]: there should be a c++11 solution, but this was the first I found

**Comment by JanMarvin** [comment]
``` r
library(openxlsx2)

# write a workbook with a few colored cells
wb <- wb_workbook()$
  add_worksheet()$
  add_fill(dims = "A1:B2;C3:D4;E1:F2;G5:H6;I2:K2", color = wb_color("yellow"))

# force a dataset into a specific cell dimension
options("openxlsx2.enforce_dims" = TRUE)
wb$add_data(dims = "I2:J2;A1:B2;G5:H6", x = matrix(1:8, 4, 2))
options("openxlsx2.enforce_dims" = FALSE)

wb$add_font(dims = "B2;G6;J2", bold = TRUE)

if (interactive()) wb$open()
```

Screenshot 2024-04-13 at 19 09 48


Not sure if this has any impact on our usual behavior. Enforce only works with data, therefore it might be a good idea to add it as option.

File: R/wb_functions.R [hunk 3]
```diff
@@ -40,46 +44,70 @@ dims_to_dataframe <- function(dims, fill = FALSE) {
...
+      fill = fill
+    )
+  } else {
+    # create data frame from rows/
+    dims_to_df2(
+      rows   = rows_out,
+      cols   = cols_out,
+      filled = filled,
+      fill   = fill
+    )
```
Comment (JanMarvin) [inline R/wb_functions.R:61]: The sort order is not necessarily correct. Should be fixed. Otherwise selecting first column name and first row name might provide incorrect results:
```R
> dims_to_dataframe("I2:J2;A1:B2;G5:H6")
     I    J    A    B    G    H
2 <NA> <NA> <NA> <NA>          
1           <NA> <NA>          
5                     <NA> <NA>
6                     <NA> <NA>
```

File: R/wb_functions.R [hunk 4]
```diff
@@ -40,46 +44,70 @@ dims_to_dataframe <- function(dims, fill = FALSE) {
...
+    out <- NULL
+    for (col in seq_along(cols)) {
+      for (row in seq_along(rows)) {
+        tmp <- paste0(
+          cols[[col]][[1]], rows[[row]][[1]],
+          ":",
+          rev(cols[[col]])[[1]],  rev(rows[[row]])[[1]]
+        )
+        out <- c(out, tmp)
+      }
```
Comment (JanMarvin) [inline R/wb_functions.R:95]: For `dim_break` this code creates incorrect results.

File: R/write.R [hunk 5]
```diff
@@ -299,7 +302,10 @@ write_data2 <- function(
     data <- as.data.frame(t(data))
   }
 
-  dims <- fits_in_dims(x = data, dims = dims, startCol = startCol, startRow = startRow)
+  # hackish solution
+  if (!getOption("openxlsx2.enforce_dims", default = FALSE)) {
+    dims <- fits_in_dims(x = data, dims = dims, startCol = startCol, startRow = startRow)
+  }
```
Comment (JanMarvin) [inline R/write.R:308]: This option is a little all over the place not a huge fan

File: R/write.R [hunk 6]
```diff
@@ -932,7 +957,7 @@ write_data_table <- function(
     na.strings      = na.strings,
     data_table      = data_table,
     inline_strings  = inline_strings,
-    dims            = dims
+    dims            = odims
```
Comment (JanMarvin) [inline R/write.R:960]: at this stage `dims` is the result of `dims_to_rowcol()` iirc. I overwrote whatever else was going in the code above. Might or might not be needed anymore. Dunno. The write-code is a bottomless pit and you better don't stare to long in it ...

File: src/helper_functions.cpp [hunk 7]
```diff
@@ -237,6 +237,76 @@ SEXP copy(SEXP x) {
...
+    std::string result;
+    for (char c : str) {
+        if (!std::isdigit(c)) {
+            result += c;
+        }
+    }
+    return result;
+}
+
+std::string rm_alf(const std::string& str) {
```
Comment (JanMarvin) [inline src/helper_functions.cpp:250]: Alf

File: src/helper_functions.cpp [hunk 8]
```diff
@@ -237,6 +237,76 @@ SEXP copy(SEXP x) {
   return Rf_duplicate(x);
 }
 
+std::string rm_dig(const std::string& str) {
```
Comment (JanMarvin) [inline src/helper_functions.cpp:240]: I lowkey believe that there are already functions to do just this ...

**Comment by JanMarvin** [comment]
@olivroy This PR is closing in, just minor things from my side. It would be nice if the disabled test would work, but at the moment I’m not in the mood to figure this out. Guess we can live without it, at least for the moment. Returning different combinations of (squared) ranges and cells is a little tricky.

But finally we can and probably should extend `wb_dims()` with this PR. We should be able to select multiple columns from a data frame, but we might be able to extend it even further. 

1) Maybe provide column and row values, but return only specific cells
```r
wb_dims(cols = 1:2, rows = c(1,3), cells = TRUE)
#> “A1,B3”
```
or 

2) highlight a dataframe based on a condition
```r
wb_dims(x = mtcars[mtcars > 4], cells = TRUE)
```

Making `wb_dims()` more flexible has been requested on [SO](https://stackoverflow.com/a/78349187/12340029)

What might be an issue is speed. I’ve tested this PR only with small dims, only a few cells not hundreds or thousands, but we’ll see…

File: R/write.R [hunk 9]
```diff
@@ -216,9 +220,16 @@ write_data2 <- function(
     na.strings = na_strings(),
     data_table = FALSE,
     inline_strings = TRUE,
-    dims = NULL
+    dims = NULL,
```
Comment (olivroy) [inline R/write.R:223]: Maybe adding dims to `write_xlsx()` could be its own PR if there are too manu moving parts. Would need to check `rows`, `cols`, `start_row` etc. ?

File: R/write.R [hunk 10]
```diff
@@ -1059,6 +1097,7 @@ write_data <- function(
     remove_cell_style = FALSE,
     na.strings        = na_strings(),
     inline_strings    = TRUE,
+    enforce           = FALSE,
```
Comment (olivroy) [inline R/write.R:1100]: Adding another argument everywhere seems a bit verbose..

I don't know if this is good, but possibly via options is better? or 

adding it to user facing function save in temp env var, then apply it where needed?

Probably not a good suggestion

**Review by olivroy — COMMENTED** [review]
Seems interesting!

So if I understand correctly, `dims = "A1;B2:C3"` didn't work before this PR?

File: R/write.R (hunk 10 above)
Comment (JanMarvin) [reply R/write.R:1100]: If I understand you correctly: I made it an argument, because with the option you have to actively set it if needed and it didn’t work with fill either way.

**Comment by JanMarvin** [comment]
> Seems interesting!
> 
> So if I understand correctly, `dims = "A1;B2:C3"` didn't work before this PR?

Yes previously this didn’t work and now it will work only if you have a scalar and a matrix with 4 values. There are obviously cornerstones that I haven’t figured out yet, but this PR is a huge step in the right direction.

**Comment by JanMarvin** [comment]
``` r
library(openxlsx2)
wb <- wb_workbook()$add_worksheet()$
  add_data(dims = "A1;B2:C3", x = 1:5, enforce = TRUE)
```

This works too, but I see the strength of different dim ranges more in evenly stretched columns:

``` r
library(openxlsx2)
wb <- wb_workbook()$
  add_worksheet()$
  add_data(dims = "A1", x = mtcars)$
  add_fill(dims = wb_dims(x = mtcars, cols = c("cyl", "wt")), color = wb_color("green"))
```

Where we previously could only color a single column, we are now able to color both. If `wb_dims(x = mtcars, cols = c("cyl", "wt"))` returns something like `"B2:B33,F2:F33"`.

File: R/write.R (hunk 9 above)
Comment (JanMarvin) [reply R/write.R:223]: Interesting, I wasn't aware that we do not support `dims` in `write_xlsx()`, but maybe that's not so important. Presumably people that use `write_xlsx()` only want a data frame or vector written as xlsx file and do not care that much about styling.

**Comment by JanMarvin** [comment]
This works:

```R
wb <- wb_workbook()$add_worksheet()$
  add_data(dims = "A1:B5", x = matrix(1:10, ncol = 2),
           enforce = TRUE, col_names = FALSE)

wb <- wb_workbook()$add_worksheet()$
  add_data(dims = "A1:B1;A3:B6", x = matrix(1:10, ncol = 2),
           enforce = TRUE, col_names = FALSE)
```

This does not work:
```R
wb <- wb_workbook()$add_worksheet()$
  add_data(dims = "A1:A5,B1:B5", x = matrix(1:10, ncol = 2),
           enforce = TRUE, col_names = FALSE)
```
